lib/authedAction.ts

```diff
@@ -0,0 +1,56 @@
+import { z } from 'zod';
+import { createServerActionProcedure } from 'zsa';
+import { requireApiAuth } from '~/lib/auth';
+import { Role } from '@prisma/client';
+import { getStudyUser } from '~/server/queries/studies';
+
+/*
+ * zsa procedure used to check if the user is authenticated. No role-based authorization.
+ * Returns the session object if the user is authenticated.
+ * session will be available in the ctx of the wrapped action.
+ */
+export const authedAction = createServerActionProcedure().handler(async () => {
+  try {
+    const session = await requireApiAuth();
+
+    return session;
+  } catch {
+    throw new Error('User not authenticated. Access denied.');
+  }
+});
+
+/*
+ * Base schema for authedActionWithRoles
+ * Can be extended with additional form data in the action definition.
+ */
+
+export const authedActionWithRolesSchema = z.object({
+  publicStudyId: z.string(),
+  roles: z.array(z.nativeEnum(Role)),
+});
+
+/*
+ * zsa procedure used to wrap other actions in role-based authorization.
+ * It returns the studyUser object if the user has the required role.
+ * studyUser will be available in the ctx of the wrapped action.
+ */
+
+export const authedActionWithRoles = createServerActionProcedure(authedAction)
+  .input(authedActionWithRolesSchema)
+  .handler(async ({ input, ctx }) => {
+    try {
+      const studyUser = await getStudyUser(ctx.userId, input.publicStudyId);
+
+      if (!studyUser) {
+        throw new Error('Role-based access denied');
+      }
+
+      if (!input.roles.includes(studyUser.role)) {
+        throw new Error('Role-based access denied');
+      }
+
+      return studyUser;
+    } catch {
+      throw new Error('Role-based access denied');
```

Comment:
Could be:  "Role-based access denied: Wrong `publicStudyId` provided or user doesn't have role-specific permission"?

Reply:
resolved - [7ceddb9](https://github.com/complexdatacollective/Studio/pull/29/commits/7ceddb9ca12ae7c2bb3441a220bbf5a43d12a00f)

fyi for this one, made it clear that there is an error checking user roles because the other two cases should be handled by the early returns above.